packages/@ember/-internals/runtime/lib/mixins/array.ts

```diff
@@ -22,17 +22,26 @@ import { ENV } from '@ember/-internals/environment';
 import Observable from '../mixins/observable';
 import MutableEnumerable from './mutable_enumerable';
 import { typeOf } from '../type-of';
+import type { AnyFn } from '@ember/-internals/utils/types';
+import type { ComputedPropertyCallback } from '@ember/-internals/metal/lib/computed';
 
-const EMPTY_ARRAY = Object.freeze([]);
+type Value<T, K extends string> = K extends keyof T ? T[K] : unknown;
 
-const identityFunction = (item) => item;
+// SAFETY: Nothing unsafe here, just TS can't infer the specificity we can.
+const EMPTY_ARRAY = Object.freeze([]) as readonly [];
 
-export function uniqBy(array, key = identityFunction) {
+const identityFunction = <T>(item: T) => item;
+
+export function uniqBy<T>(
+  array: T[] | EmberArray<T>,
+  keyOrFunc: string | ((item: T) => unknown) = identityFunction
+): T[] | EmberArray<T> {
   assert(`first argument passed to \`uniqBy\` should be array`, isArray(array));
 
-  let ret = A();
+  // SAFETY: This created an empty array so casting it to our known type is safe.
+  let ret = A() as NativeArray<T>;
```

Comment:
Huh. Is `A()` not generic?

Reply:
I refactored the types for `A` later and missed this.